File: kuwala_core/src/pricing.rs
const INV_SQRT_2PI: f64 = 0.398942280401432677939946059934;

/// Standard normal cumulative distribution function (Abramowitz and Stegun approximation with high accuracy)
#[inline]
pub fn norm_cdf(x: f64) -> f64 {
    if x.is_nan() {
        return f64::NAN;
    }
    // Erf based standard normal CDF
    0.5 * (1.0 + erf(x / std::f64::consts::SQRT_2))
}

/// Standard normal probability density function
#[inline]
pub fn norm_pdf(x: f64) -> f64 {
    INV_SQRT_2PI * (-0.5 * x * x).exp()
}

/// Error function approximation (Chebyshev fitting with precision ~ 1.5e-7 or exact erf)
#[inline]
pub fn erf(x: f64) -> f64 {
    // Standard numerical approximation for erf(x) with maximum error 1.2e-7
    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let x_abs = x.abs();

    // Coefficients
    let a1 = 0.254829592;
    let a2 = -0.284496736;
    let a3 = 1.421413741;
    let a4 = -1.453152027;
    let a5 = 1.061405429;
    let p = 0.3275911;

    let t = 1.0 / (1.0 + p * x_abs);
    let poly = ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t;
    let y = 1.0 - poly * (-x_abs * x_abs).exp();

    sign * y
}

/// Black-Scholes European option pricing
/// spot: Current spot price S
/// strike: Strike price K
/// t: Time to expiration in years T
/// r: Risk-free interest rate
/// q: Continuous dividend yield / borrow cost
/// sigma: Volatility
/// is_call: true for Call, false for Put
pub fn black_scholes_price(
    spot: f64,
    strike: f64,
    t: f64,
    r: f64,
    q: f64,
    sigma: f64,
    is_call: bool,
) -> f64 {
    if t <= 0.0 {
        return if is_call {
            (spot - strike).max(0.0)
        } else {
            (strike - spot).max(0.0)
        };
    }
    if sigma <= 0.0 {
        let forward = spot * ((r - q) * t).exp();
        let discount = (-r * t).exp();
        return if is_call {
            discount * (forward - strike).max(0.0)
        } else {
            discount * (strike - forward).max(0.0)
        };
    }

    let sqrt_t = t.sqrt();
    let vol_sqrt_t = sigma * sqrt_t;
    let d1 = ((spot / strike).ln() + (r - q + 0.5 * sigma * sigma) * t) / vol_sqrt_t;
    let d2 = d1 - vol_sqrt_t;

    let df_r = (-r * t).exp();
    let df_q = (-q * t).exp();

    if is_call {
        spot * df_q * norm_cdf(d1) - strike * df_r * norm_cdf(d2)
    } else {
        strike * df_r * norm_cdf(-d2) - spot * df_q * norm_cdf(-d1)
    }
}

/// Black-76 European futures/forward option pricing
pub fn black76_price(
    forward: f64,
    strike: f64,
    t: f64,
    r: f64,
    sigma: f64,
    is_call: bool,
) -> f64 {
    if t <= 0.0 {
        return if is_call {
            (forward - strike).max(0.0)
        } else {
            (strike - forward).max(0.0)
        };
    }
    if sigma <= 0.0 {
        let discount = (-r * t).exp();
        return if is_call {
            discount * (forward - strike).max(0.0)
        } else {
            discount * (strike - forward).max(0.0)
        };
    }

    let sqrt_t = t.sqrt();
    let vol_sqrt_t = sigma * sqrt_t;
    let d1 = ((forward / strike).ln() + 0.5 * sigma * sigma * t) / vol_sqrt_t;
    let d2 = d1 - vol_sqrt_t;
    let discount = (-r * t).exp();

    if is_call {
        discount * (forward * norm_cdf(d1) - strike * norm_cdf(d2))
    } else {
        discount * (strike * norm_cdf(-d2) - forward * norm_cdf(-d1))
    }
}

File: julia/src/pricing.jl
# Kuwala Julia Analytical Pricing Module
# High-precision Black-Scholes and Black-76 options pricer

const INV_SQRT_2PI = 0.39894228040143267793994605993438
const SQRT_2 = 1.4142135623730950488016887242097

"""
    norm_cdf(x::Real) -> Float64
High-precision standard normal cumulative distribution function.
"""
function norm_cdf(x::Real)
    # Rational Chebyshev approximation / erfc formula
    return 0.5 * (1.0 + erf_approx(x / SQRT_2))
end

"""
    erf_approx(x::Real) -> Float64
Chebyshev rational approximation to the error function with error < 1.2e-7.
"""
function erf_approx(x::Real)
    # Standard Abramowitz & Stegun formula 7.1.26
    sign_x = sign(x)
    abs_x = abs(Float64(x))
    t = 1.0 / (1.0 + 0.3275911 * abs_x)
    poly = t * (0.254829592 + t * (-0.284496736 + t * (1.421413741 + t * (-1.453152027 + t * 1.061405429))))
    return sign_x * (1.0 - poly * exp(-abs_x * abs_x))
end

"""
    norm_pdf(x::Real) -> Float64
Standard normal probability density function.
"""
function norm_pdf(x::Real)
    return INV_SQRT_2PI * exp(-0.5 * Float64(x)^2)
end

"""
    black_scholes(spot, strike, t, r, q, sigma; is_call=true) -> Float64
Analytical European Option pricing formula under the Black-Scholes model.
"""
function black_scholes(
    spot::Real,
    strike::Real,
    t::Real,
    r::Real,
    q::Real,
    sigma::Real;
    is_call::Bool = true
)::Float64
    s = Float64(spot)
    k = Float64(strike)
    ttm = Float64(t)
    rate = Float64(r)
    div = Float64(q)
    vol = Float64(sigma)

    if ttm <= 0.0
        return is_call ? max(0.0, s - k) : max(0.0, k - s)
    end
    if vol <= 0.0
        df_r = exp(-rate * ttm)
        df_q = exp(-div * ttm)
        return is_call ? max(0.0, s * df_q - k * df_r) : max(0.0, k * df_r - s * df_q)
    end
    if s <= 0.0
        return is_call ? 0.0 : k * exp(-rate * ttm)
    end
    if k <= 0.0
        return is_call ? s * exp(-div * ttm) : 0.0
    end

    df_r = exp(-rate * ttm)
    df_q = exp(-div * ttm)
    sqrt_t = sqrt(ttm)
    d1 = (log(s / k) + (rate - div + 0.5 * vol * vol) * ttm) / (vol * sqrt_t)
    d2 = d1 - vol * sqrt_t

    if is_call
        return s * df_q * norm_cdf(d1) - k * df_r * norm_cdf(d2)
    else
        return k * df_r * norm_cdf(-d2) - s * df_q * norm_cdf(-d1)
    end
end

"""
    black76(forward, strike, t, r, sigma; is_call=true) -> Float64
Analytical Black-76 European Commodity / Futures option pricer.
"""
function black76(
    forward::Real,
    strike::Real,
    t::Real,
    r::Real,
    sigma::Real;
    is_call::Bool = true
)::Float64
    f = Float64(forward)
    k = Float64(strike)
    ttm = Float64(t)
    rate = Float64(r)
    vol = Float64(sigma)

    if ttm <= 0.0
        return is_call ? max(0.0, f - k) : max(0.0, k - f)
    end
    df = exp(-rate * ttm)
    if vol <= 0.0
        return df * (is_call ? max(0.0, f - k) : max(0.0, k - f))
    end

    sqrt_t = sqrt(ttm)
    d1 = (log(f / k) + 0.5 * vol * vol * ttm) / (vol * sqrt_t)
    d2 = d1 - vol * sqrt_t

    if is_call
        return df * (f * norm_cdf(d1) - k * norm_cdf(d2))
    else
        return df * (k * norm_cdf(-d2) - f * norm_cdf(-d1))
    end
end
